src/redisearch_rs/c_entrypoint/iterators_ffi/src/c2rust.rs

```diff
@@ -0,0 +1,309 @@
+    /// 2. `header` is an owning pointer, in the same way `Box` owns the
+    ///    allocated heap data.
+    pub(crate) const unsafe fn new(header: NonNull<QueryIterator>) -> Self {
+        // SAFETY: the caller is required to uphold `Self::header` field invariants.
+        Self { header }
+    }
+
+    /// Return a raw pointer to the underlying [`QueryIterator`].
+    ///
+    /// The caller is taking ownership of the [`QueryIterator`] instance and is
+    /// therefore responsible to free its contents.
+    pub(crate) fn into_raw(self) -> NonNull<QueryIterator> {
+        let self_ = ManuallyDrop::new(self);
+        self_.header
+    }
+}
+
+impl<'index> RQEIterator<'index> for CRQEIterator {
+    fn read(&mut self) -> Result<Option<&mut RSIndexResult<'index>>, RQEIteratorError> {
+        let callback = self
+            .Read
+            .expect("The `Read` callback is a NULL function pointer");
+        // SAFETY:
+        // - We have a unique handle over this iterator.
+        // - The C code must guarantee, by constructor, that callbacks
+        //   can be called on types that implement its C iterator API.
+        let status = unsafe { callback(self.header.as_ptr()) };
+        #[allow(non_upper_case_globals)]
+        match status {
+            IteratorStatus_ITERATOR_EOF => {
+                self.atEOF = true;
+                Ok(None)
+            }
+            IteratorStatus_ITERATOR_TIMEOUT => Err(RQEIteratorError::TimedOut),
+            IteratorStatus_ITERATOR_OK => {
+                let data = self.current as *mut RSIndexResult;
+                // SAFETY:
+                // - We have a unique handle over this iterator.
+                let data = unsafe {
+                    data.as_mut()
+                        .expect("`current` is a NULL pointer after `Read` returned OK")
+                };
+                Ok(Some(data))
+            }
+            IteratorStatus_ITERATOR_NOTFOUND => {
+                unreachable!(
+                    "The `Read` callback returned `NOTFOUND`, which should only be used for `SkipTo`"
+                )
+            }
+            _ => {
+                unreachable!("`Read` returned an unexpected iterator status, {status}")
+            }
+        }
+    }
+
+    fn skip_to(
+        &mut self,
+        doc_id: t_docId,
+    ) -> Result<Option<SkipToOutcome<'_, 'index>>, RQEIteratorError> {
+        let callback = self
+            .SkipTo
+            .expect("The `SkipTo` callback is a NULL function pointer");
+        // SAFETY:
+        // - We have a unique handle over this iterator.
+        // - The C code must guarantee, by constructor, that callbacks
+        //   can be called on types that implement its C iterator API.
+        let status = unsafe { callback(self.header.as_ptr(), doc_id) };
+        #[allow(non_upper_case_globals)]
+        match status {
+            IteratorStatus_ITERATOR_EOF => {
+                self.atEOF = true;
```

Comment:
why do we need to set this value? That is something the C QueryIterator would already set if desired, no?